### ISSUE TITLE
Path helper picks up surrounding scope only if the URL has a parameter

### ISSUE THREAD

**Issue by manueljacob** [rails/rails#33219]
Given this routes configuration:

```
Rails.application.routes.draw do
  scope "(:locale)", locale: /en|de/ do
    get 'test', to: 'application#test'
    get 'with_param/:foo', to: 'application#with_param', as: 'with_param'
    get 'without_param', to: 'application#without_param'
  end
end
```

When GETting `http://localhost:3000/test`:

* `with_param_path(foo: 'bar')` returns `/with_param/bar` 
* `without_param_path` returns `/without_param?foo=bar`.

When GETting `http://localhost:3000/en/test`:

* `with_param_path(foo: 'bar')` returns `/en/with_param/bar`
* `without_param_path` returns `/without_param`.

This seems inconsistent to me. I would expect that the path helper includes the locale into the resulting URL if and only if it's present in the current URL, independent of whether the new URL contains a parameter or not.

**Rails version**: 5.2.0
**Ruby version**: 2.5.1p57

**Comment by manueljacob** [comment]
The relevant code is in https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/journey/formatter.rb#L69

When GETting `http://localhost:3000/en/test`, when calculating `with_param_path(foo: 'bar')`,

* `options` is `{:controller=>"application", :action=>"with_param", :foo=>"bar"}`
* `recall` is `{:controller=>"application", :action=>"test", :locale=>"en"}`
* `route.parts` is `locale`, `foo`, `format`
* `route.parts.reverse_each.drop_while { |part| !options.key?(part) || (options[part] || recall[part]).nil? }` is `foo`, `locale`
* `route.required_parts` is `foo`

When calculating `without_param_path`,
* `options` is `{:controller=>"application", :action=>"without_param"}`
* `recall` is `{:controller=>"application", :action=>"test", :locale=>"en"}`
* `route.parts` is `locale`, `format`
* `route.parts.reverse_each.drop_while { |part| !options.key?(part) || (options[part] || recall[part]).nil? }` is empty
* `route.required_parts` is empty

**Comment by manueljacob** [comment]
*ping*

Is there any additional information I could provide to help resolving this issue?

**Comment by sikachu** [comment]
This definitely seems like a bug to me, given in the second scenario `locale` is also got set in the url itself.

Would you like to work on writing a failing test case and submit a patch to fix this issue?

/cc @tenderlove if you have some time to take a look.

**Comment by manueljacob** [comment]
Yes, I'd like to look into this and at least write a test case.

However I'm unsure where to put the tests. In one of these?

- actionpack/test/dispatch/routing_test.rb
- actionpack/test/controller/routing_test.rb
- actionpack/test/journey/router_test.rb

How can I CC someone? Or should I just mention @tenderlove ?

**Comment by gmcgibbon** [comment]
I can't seem to reproduce this issue using 6.0 or 5.2.0 with the code provided. Can you provide a failing test script ([template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/action_controller_master.rb)) or a sample app?

**Comment by albertoalmagro** [comment]
I was able to reproduce it with the following reproduction steps:
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    scope "(:locale)", locale: /en|de/ do
      get 'test', to: 'application#test'
      get 'with_param/:foo', to: 'application#with_param', as: 'with_param'
      get 'without_param', to: 'application#without_param'
    end
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def test
    render plain: "Home"
  end

  def with_param; end
  def without_param; end
end

require "minitest/autorun"
require "rack/test"

class BugTest < Minitest::Test
  include Rack::Test::Methods
  include Rails.application.routes.url_helpers

  def test_returns_success
    get "/test"
    assert_equal "/with_param/bar", with_param_path(foo: 'bar')
    assert_equal "/without_param", without_param_path
  end

  def test_returns_error
    get "/en/test" do
      assert_equal "/en/with_param/bar", with_param_path(foo: 'bar')
      assert_equal "/en/without_param", without_param_path
    end
  end

  private
    def app
      Rails.application
    end
end
```

May I submit a PR to fix it?

**Comment by albertoalmagro** [comment]
I could reproduce it with the following reproduction steps:
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    scope "(:locale)", locale: /en|de/ do
      get 'test', to: 'test#test'
      get 'with_param/:foo', to: 'test#with_param', as: 'with_param'
      get 'without_param', to: 'test#without_param'
    end
  end
end

module PathBuffer
  class << self
    def clear
      values.clear
    end

    def add(value)
      values << value
    end

    def values
      @values ||= []
    end

    def with_param
      values.first
    end

    def without_param
      values.second
    end
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def test
    PathBuffer.add with_param_path(foo: 'bar')
    PathBuffer.add without_param_path
  end

  def with_param; end
  def without_param; end
end

require "minitest/autorun"
require "rack/test"

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def setup
    PathBuffer.clear
  end

  def test_returns_success
    get "/test"

    assert_equal "/with_param/bar", PathBuffer.with_param
    assert_equal "/without_param", PathBuffer.without_param
  end

  def test_returns_error
    get "/en/test"

    assert_equal "/en/with_param/bar", PathBuffer.with_param
    assert_equal "/en/without_param", PathBuffer.without_param
  end

  private
    def app
      Rails.application
    end
end
```
May I look further into it and submit a PR to fix it?

Thanks in advance.

**Comment by gmcgibbon** [comment]
@albertoalmagro Nice work! Please feel free to PR a fix.

**Comment by albertoalmagro** [comment]
This isn't stale, it has been approved and it's waiting to be merged #34656

**Comment by andreibondarev** [comment]
Is it fair to suggest that this behavior, raised by this issue and fixed, should be applicable to engine routes as well? Example:

```
Rails.application.routes.draw do
  scope "(:locale)", locale: /en|de/ do
   mount Thredded::Engine => '/forum'
```

I would imagine that starting out from `/en/forum`, all internal url helpers inside the engine ([Thredded](https://github.com/thredded/thredded) in this case) would utilize `locale: :en` when building url.

Unfortunately I'm unable to get it to work. Does anyone have any insight into this by chance?
